--- IscStatement.cpp.orig	2006-04-05 19:09:13.000000000 +0200
+++ IscStatement.cpp	2007-07-18 19:39:12.000000000 +0200
@@ -298,11 +298,13 @@
 		typeStmt = stmtNone;
 		if ( connection )
 		{
+			USHORT option = 0;
 			if (connection->autoCommit)
 				connection->commitAuto();
-			// Close cursors too.
+			else
+				option = DSQL_close; // Close cursors too.
 			ISC_STATUS statusVector [20];
-			connection->GDS->_dsql_free_statement (statusVector, &statementHandle, DSQL_close);
+			connection->GDS->_dsql_free_statement (statusVector, &statementHandle, option);
 			// Cursor already closed or not assigned
 			if ( statusVector[1] && statusVector[1] != 335544569 )
 				THROW_ISC_EXCEPTION (connection, statusVector);
