Index: E:/FbOdbc/IscDbc/MultibyteConvert.cpp
===================================================================
--- E:/FbOdbc/IscDbc/MultibyteConvert.cpp	(revision 9)
+++ E:/FbOdbc/IscDbc/MultibyteConvert.cpp	(revision 10)
@@ -594,10 +594,10 @@
 				UChar32 c2;
 
 				if ( U_IS_SURROGATE_LEAD( c ) 
-					&& wcsOrg < wcsEnd
-					&& U16_IS_TRAIL( c2 = *wcsOrg ) )
+					&& wcsOrg + i < wcsEnd
+					&& U16_IS_TRAIL( c2 = wcsOrg[i] ) )
 				{
-					++wcsOrg;
+					++i;
 					c = U16_GET_SUPPLEMENTARY( c, c2 );
 				}
 				else
