From 6978aab7d596ae9f57405be05ec4b062cd3961eb Mon Sep 17 00:00:00 2001
From: Alexpux <alexey.pawlow@gmail.com>
Date: Thu, 20 Nov 2014 07:42:19 +0300
Subject: [PATCH] mingw-w64 don't need hack for SecPkgContext_AccessToken

---
 src/auth/trusted/AuthSspi.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/auth/trusted/AuthSspi.cpp b/src/auth/trusted/AuthSspi.cpp
index 9103bdb..4174a54 100644
--- a/src/auth/trusted/AuthSspi.cpp
+++ b/src/auth/trusted/AuthSspi.cpp
@@ -135,7 +135,7 @@ AuthSspi::~AuthSspi()
 
 bool AuthSspi::checkAdminPrivilege(PCtxtHandle phContext) const
 {
-#if defined (__GNUC__)
+#if defined (__GNUC__) && !defined(__MINGW64_VERSION_MAJOR)
 	// ASF: MinGW hack.
 	struct SecPkgContext_AccessToken
 	{
-- 
2.1.3

