Index: src/jrd/isc_sync.cpp
===================================================================
--- src/jrd/isc_sync.cpp	(revision 51530)
+++ src/jrd/isc_sync.cpp	(working copy)
@@ -320,8 +320,8 @@
 	public:
 		const static int N_FILES = 8;
 		const static int N_SETS = 256;
-#ifdef DEV_BUILD
-		const static int SEM_PER_SET = 4;	// force multiple sets allocation
+#if defined(DEV_BUILD) || defined(FREEBSD)
+		const static int SEM_PER_SET = 4;	// force multiple sets allocation || work with default freebsd kernel
 #else
 		const static int SEM_PER_SET = 31;	// hard limit for some old systems, might set to 32
 #endif
