--- f:\tmp\ori\parse.y	Wed Aug  1 14:09:26 2007
+++ f:\tmp\new\parse.y	Wed Aug  1 14:04:42 2007
@@ -516,6 +516,7 @@ static LexerState lex;
 %token ABS
 %token ACCENT
 %token ACOS
+%token ALWAYS
 %token ASCII_CHAR
 %token ASCII_VAL
 %token ASIN
@@ -538,6 +539,7 @@ static LexerState lex;
 %token EXP
 %token FLOOR
 %token GEN_UUID
+%token GENERATED
 %token GLOBAL 
 %token HASH
 %token INSENSITIVE
@@ -1336,12 +1338,16 @@ column_def	: column_def_name data_type_o
  * source text
  */
 
-def_computed	: computed_by '(' begin_trigger value end_trigger ')'
+def_computed	: generation_clause '(' begin_trigger value end_trigger ')'
 			{ 
 			lex.g_field->fld_flags |= FLD_computed;
 			$$ = make_node (nod_def_computed, 2, $4, $5); }
 		;
 
+generation_clause :	GENERATED ALWAYS AS
+		| computed_by
+ 		;
+
 computed_by	: COMPUTED BY
 		| COMPUTED
 		;
@@ -4746,6 +4752,7 @@ non_reserved_word :
 	| ABS					/* added in FB 2.1 */
 	| ACCENT
 	| ACOS
+	| ALWAYS
 	| ASCII_CHAR
 	| ASCII_VAL
 	| ASIN
@@ -4766,6 +4773,7 @@ non_reserved_word :
 	| EXP
 	| FLOOR
 	| GEN_UUID
+	| GENERATED
 	| HASH
 	| LIST
 	| LN
