--- f:\tmp\ori\keywords.cpp	Wed Aug  1 14:09:44 2007
+++ f:\tmp\new\keywords.cpp	Wed Aug  1 13:46:18 2007
@@ -62,6 +62,7 @@ static const TOK tokens[] =
 	{AFTER, "AFTER", 1, false},
 	{ALL, "ALL", 1, false},
 	{ALTER, "ALTER", 1, false},
+	{ALWAYS, "ALWAYS", 2, false},
 	{AND, "AND", 1, false},
 	{ANY, "ANY", 1, false},
 	{AS, "AS", 1, false},
@@ -177,6 +178,7 @@ static const TOK tokens[] =
 	{FULL, "FULL", 1, false},
 	{FUNCTION, "FUNCTION", 1, false},
 	{GDSCODE, "GDSCODE", 1, false},
+	{GENERATED, "GENERATED", 2, false},
 	{GENERATOR, "GENERATOR", 1, false},
 	{GEN_ID, "GEN_ID", 1, false},
 	{GEN_UUID, "GEN_UUID", 2, false},
