cvs diff -u -- OdbcStatement.cpp (in directory D:\Sources\OdbcFb\OdbcJdbc)
Index: OdbcStatement.cpp
===================================================================
RCS file: /cvsroot/firebird/OdbcJdbc/OdbcStatement.cpp,v
retrieving revision 1.78
diff -u -r1.78 OdbcStatement.cpp
--- OdbcStatement.cpp	4 Aug 2012 10:55:00 -0000	1.78
+++ OdbcStatement.cpp	15 Oct 2012 09:33:24 -0000
@@ -1480,8 +1480,14 @@
 
 	if ( crowKeyset > crowRowset )
 		sqlSetStmtAttr(SQL_ATTR_CURSOR_TYPE, (SQLPOINTER)SQL_CURSOR_KEYSET_DRIVEN, 0);
-	else
-		sqlSetStmtAttr(SQL_ATTR_CURSOR_TYPE, crowKeyset < 0 ? (SQLPOINTER)-crowKeyset : (SQLPOINTER)crowKeyset, 0);
+	else {
+		SQLPOINTER ptr;
+		if (crowKeyset < 0)
+			ptr = (SQLPOINTER)-crowKeyset;
+		else
+			ptr = (SQLPOINTER)crowKeyset;
+		sqlSetStmtAttr(SQL_ATTR_CURSOR_TYPE, ptr, 0);
+	}
 
 	sqlSetStmtAttr(SQL_ATTR_CONCURRENCY, (SQLPOINTER)(intptr_t)fConcurrency, 0);
 
