diff -urN Firebird-2.5.8.27089-0.orig/builds/posix/darwin.defaults Firebird-2.5.8.27089-0/builds/posix/darwin.defaults
--- Firebird-2.5.8.27089-0.orig/builds/posix/darwin.defaults	2018-01-05 20:38:19.000000000 +1100
+++ Firebird-2.5.8.27089-0/builds/posix/darwin.defaults	2018-07-10 11:47:51.000000000 +1000
@@ -53,7 +53,7 @@
 LINK_IB_UTIL = $(LIB_LINK) $(LINK_IBUTIL_SYMBOLS) $(LIB_LINK_OPTIONS)\
 $(LIB_LINK_RPATH) $(LIB_LINK_SONAME)
 
-LINK_INTL = $(LIB_LINK) $(LIB_BUNDLE_OPTIONS)
+LINK_INTL = $(LIB_LINK) $(LINK_FBINTL_SYMBOLS) $(LIB_BUNDLE_OPTIONS)
 
 LINK_TRACE = $(LIB_LINK) $(LIB_BUNDLE_OPTIONS)
 
